### PR TITLE
feat(incidents): #MA-949 add class/groups searchbar on incidents list

### PR DESCRIPTION
**Pull request by flmartineau**
## Describe your changes

- Add a class/groups search bar for the incidents list

## Checklist tests

- check if searching classes is working properly
- check if searching students is still behaving normally

## Issue ticket number and link

https://entsupport.gdapublic.fr/browse/MA-949

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java [hunk 1]
```diff
@@ -29,15 +30,18 @@
...
 public class IncidentsController extends ControllerHelper {
 
     private IncidentsService incidentsService;
+    private GroupService groupService;
     private EventBus eb;
 
     public IncidentsController(EventBus eb) {
         super();
         this.incidentsService = new DefaultIncidentsService(eb);
+        this.groupService = new DefaultGroupService(eb);
```
Comment (JinDouang) [inline incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java]: CommonService!

File: incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java [hunk 2]
```diff
@@ -65,79 +69,103 @@ public void view(HttpServerRequest request) {
...
+                .onSuccess(groupStudents -> {
+                    Promise<JsonArray> incidentsPromise = Promise.promise();
+                    Promise<JsonObject> pageNumberPromise = Promise.promise();
 
-        incidentsService.get(structureId, startDate, endDate,
-                userId, page, false, field, reverse, FutureHelper.handlerJsonArray(incidentsFuture));
-        incidentsService.getPageNumber(structureId, startDate, endDate, userId,
-                page, field, reverse, FutureHelper.handlerJsonObject(pageNumberFuture));
+                    List<String> studentIds = ((List<JsonObject>) groupStudents.getList()).stream()
+                            .map((JsonObject s) -> s.getString(Field.ID)).collect(Collectors.toList());
```
Comment (JinDouang) [inline incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java:94]: Si jamais tu obtiens [id, id, null] tu risques d'avoir un problème
Il faudrait le gérer :  

`studentIds .removeAll(Collections.singletonList(null))`;

File: incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java [hunk 3]
```diff
@@ -65,79 +69,103 @@ public void view(HttpServerRequest request) {
...
-                        String message = String.format("[Incidents@%s::exportIncidents] Failed to fetch incidents",
-                                this.getClass().getSimpleName());
-                        log.error(message, event.left().getValue());
-                        renderError(request);
-                        return;
-                    }
+        groupService.getGroupStudents(audienceIds)
+                .onFailure(fail -> renderError(request, JsonObject.mapFrom(fail.getCause().getMessage())))
+                .onSuccess(groupStudents -> {
+                    List<String> studentIds = ((List<JsonObject>) groupStudents.getList()).stream()
```
Comment (JinDouang) [inline incidents/src/main/java/fr/openent/incidents/controller/IncidentsController.java:137]: Si jamais tu obtiens [id, id, null] tu risques d'avoir un problème
Il faudrait le gérer :

`studentIds .removeAll(Collections.singletonList(null));`

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: incidents/src/main/resources/public/ts/controllers/incidents.ts [hunk 4]
```diff
@@ -120,7 +133,9 @@ export const incidentsController = ng.controller('IncidentsController',
                 if (student && !_.find(vm.filter.students, student)) {
                     vm.filter.students.push(student);
                 }
-                setStudentToSync();
+                vm.filter.groups = vm.groupsSearch.getSelectedGroups().map(group => group["id"]);
```
Comment (BastienBenard) [inline incidents/src/main/resources/public/ts/controllers/incidents.ts:136]: Est ce que c'est possible d'essayer de typer getSelectedGroups() ?

**Review by BastienBenard — COMMENTED** [review]
Il me semble que tu n'es pas a jour avec dev

File: incidents/src/main/java/fr/openent/incidents/Incidents.java [hunk 5]
```diff
@@ -31,13 +32,14 @@ public void start() throws Exception {
         super.start();
         dbSchema = config.getString("db-schema");
         final EventBus eb = getEventBus(vertx);
+        CommonServiceFactory commonServiceFactory = new CommonServiceFactory(vertx);
```
Comment (JinDouang) [inline incidents/src/main/java/fr/openent/incidents/Incidents.java]: commonIncidentService du coup!

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: incidents/src/main/resources/public/ts/controllers/incidents.ts (hunk 4 above)
Comment (flmartineau) [reply incidents/src/main/resources/public/ts/controllers/incidents.ts:136]: Ca implique de typer tout le fichier groupSearch et potentiellement de gérer les problèmes de typages partout où il est utilisé

**Review by JinDouang — APPROVED** [review]

**Review by BastienBenard — APPROVED** [review]

File: incidents/src/main/java/fr/openent/incidents/service/CommonIncidentsServiceFactory.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import fr.openent.incidents.service.impl.*;
+import fr.openent.presences.common.service.*;
+import fr.openent.presences.common.service.impl.*;
+import io.vertx.core.*;
+import io.vertx.core.eventbus.*;
+
+public class CommonIncidentsServiceFactory {
+    private final Vertx vertx;
+
+    public CommonIncidentsServiceFactory(Vertx vertx) {
```
Comment (aymericdevriesere) [inline incidents/src/main/java/fr/openent/incidents/service/CommonIncidentsServiceFactory.java:12]: Pourquoi créer un classe dont le seul but vas être d'instancier une autre ?

**Pull request merged**